python/ray/serve/_private/replica.py

```diff
@@ -784,17 +831,36 @@ async def initialize_callable(self):
             if isinstance(self._callable, ASGIAppReplicaWrapper):
                 await self._callable._run_asgi_lifespan_startup()
 
+        self._user_health_check = getattr(self._callable, HEALTH_CHECK_METHOD, None)
+
         logger.info(
             "Finished initializing replica.",
             extra={"log_to_stderr": False},
         )
 
+    @_run_on_user_code_event_loop
+    async def _call_user_health_check(self):
+        await self._call_func_or_gen(self._user_health_check)
+
     async def call_user_health_check(self):
-        user_health_check = getattr(self._callable, HEALTH_CHECK_METHOD, None)
-        if callable(user_health_check):
-            await self._call_func_or_gen(user_health_check)
+        if self._callable is None:
+            raise RuntimeError(
+                "initialize_callable must be called before health check."
+            )
 
+        # If the user provided a health check, call it on the user code thread. If user
+        # code blocks the event loop the health check may time out.
+        #
+        # To avoid this issue for basic cases without a user-defined health check, skip
+        # interacting with the user callable entirely.
+        if self._user_health_check is not None:
+            return await self._call_user_health_check()
+
+    @_run_on_user_code_event_loop
     async def call_reconfigure(self, user_config: Any):
+        if self._callable is None:
+            raise RuntimeError("initialize_callable must be called before reconfigure.")
```

Comment:
Can extract to a method (like `_assert_initialized`)

Reply:
done